### ISSUE TITLE
Aggregate repo exclusion files in the VMR

### ISSUE THREAD

**Issue by ViktorHofer** [dotnet/dotnet#1014]
Some repositories will continue to have their own official builds and hence need to keep their exclusion files inside their repository for tools like CredScan. The VMR itself also runs these tools and needs an aggregation of all the exclusion files per tool.

This might just be an intermediate that could be produced in YML or as part of the repo build. Whatever solution we choose, the output must be available before the tool that needs it, runs.

Also see https://github.com/dotnet/arcade-services/issues/2362

**Comment by akoeplinger** [comment]
For repos that won't have an official build we should move the suppression files into the VMR to simplify things so we only need to do this for the small subset of repos which continue to have their own official build.

**Comment by ViktorHofer** [comment]
_Replied to that in chat:_

I wouldn't necessarily agree with that. If I make a change in repo A and I know that this will require an exclusion, then I want to be able to add that exclusion as part of the original change. That should be possible inside the repository.
 
Similarly, we're considering to allow repositories to define the set of allowed binaries inside the repo via an exclusion file. At the moment that binaries exclusion file is centrally defined in the VMR.
 
In general the VMR requires a repository contract. @mmitche is documenting that contract. I would add a section to that doc defining the contract for exclusion files.

**Comment by mmitche** [comment]
Ideally we would get some kind of checking at PR time in an individual repo. Whether that is possible in public PRs is an open question at this point., but it would be great to have PR time checking and then have those exclusions/suppressions apply seamlessly in the VMR build.

I could see lots of ways this could be done, from custom automation at sync time to some kind of dynamic creation of the suppression info at build time.

**Comment by akoeplinger** [comment]
> some kind of dynamic creation of the suppression info at build time.

This doesn't seem viable given that the scanners run in the `SDL Sources Analysis (self)` stage which just checks out the repo sources so we can't dynamically produce it there. It probably needs to happen during the VMR sync.

**Comment by ViktorHofer** [comment]
@akoeplinger do you know if this is already implemented. I saw the linked PR.

**Comment by akoeplinger** [comment]
It's implemented for CredScan by the above PR. There might be other cases like SignCheckExclusionsFile.txt where we could do the same.

That said, for SignCheck we can probably consolidate the files into the VMR since all of these repos probably won't have a separate official build:

<img width="664" alt="Image" src="https://github.com/user-attachments/assets/52254eb2-5224-4af8-a732-9884d534eaff" />

**Comment by mmitche** [comment]
/cc @ellahathaway

IIRC there were some challenges with automatically aggregating these at scan time. The patterns used in individual repos may not be specific enough to avoid applying to things they shouldn't in the full VMR build. It turned out, at least for signcheck, though, that the aggregated exclusion list is quite small and generic.

**Comment by ellahathaway** [comment]
> /cc [@ellahathaway](https://github.com/ellahathaway)
> 
> IIRC there were some challenges with automatically aggregating these at scan time. The patterns used in individual repos may not be specific enough to avoid applying to things they shouldn't in the full VMR build. It turned out, at least for signcheck, though, that the aggregated exclusion list is quite small and generic.

This is correct. The exclusions at the repo-level are not specific-enough to aggregate into one file at scan time. For example, if aspnetcore includes `*.js`, and sdk includes `dotnet*.js`, aggregating these exclusions would result in every .js file being skipped. The flipside to this is that we cannot make them more specific. One reason is that artifacts may shift around or change names which can lead to false positives in the VMR which is not very UX friendly.

The solution that Matt and I came up with was to just have a separate VMR exclusions file. The exclusion list is short, so it's not a big maintenance to have this extra file.

**Comment by akoeplinger** [comment]
> The solution that Matt and I came up with was to just have a separate VMR exclusions file. The exclusion list is short, so it's not a big maintenance to have this extra file.

yeah I came to the same conclusion and once runtime/aspnetcore/emsdk don't have an official build anymore we can remove the SignCheckExclusions.txt there and use just the one in the VMR

**Comment by mmitche** [comment]
Completed for the cred-scan and signcheck exclusions.
